privacyidea/lib/tokens/pushtoken.py

```diff
@@ -0,0 +1,578 @@
+                raise ParameterError("Unknown Firebase configuration!")
+            fb_options = firebase_configs[0].option_dict
+            for k in [FIREBASE_CONFIG.PROJECT_NUMBER, FIREBASE_CONFIG.PROJECT_ID,
+                      FIREBASE_CONFIG.APP_ID, FIREBASE_CONFIG.API_KEY]:
+                extra_data[k] = fb_options.get(k)
+            # We display this during the first enrollment step!
+            qr_url = create_push_token_url(url=fb_options.get(FIREBASE_CONFIG.REGISTRATION_URL),
+                                           user=user.login,
+                                           realm=user.realm,
+                                           serial=self.get_serial(),
+                                           tokenlabel=tokenlabel,
+                                           issuer=tokenissuer,
+                                           user_obj=user,
+                                           extra_data=extra_data,
+                                           ttl=fb_options.get(FIREBASE_CONFIG.TTL))
+            response_detail["pushurl"] = {"description": _("URL for privacyIDEA Push Token"),
+                                          "value": qr_url,
+                                          "img": create_img(qr_url, width=250)
+                                          }
+            self.add_tokeninfo(FIREBASE_CONFIG.PROJECT_ID, fb_options.get(FIREBASE_CONFIG.PROJECT_ID))
+
+            response_detail["enrollment_credential"] = self.get_tokeninfo("enrollment_credential")
+
+        elif self.token.rollout_state == "enrolled":
+            # in the second enrollment step we return the public key of the server to the smartphone.
+            pubkey = _strip_key(self.get_tokeninfo(PUBLIC_KEY_SERVER))
+            response_detail["public_key"] = pubkey
+
+        return response_detail
+
+    @classmethod
+    def api_endpoint(cls, request, g):
+        """
+        This provides a function which is called by the API endpoint
+        /ttype/push which is defined in api/ttype.py
+
+        The method returns
+            return "json", {}
+
+        This endpoint is used for the 2nd enrollment step of the smartphone.
+        Parameters sent:
+            * serial
+            * fbtoken
+            * pubkey
+
+        This endpoint is also used, if the smartphone sends the signed response
+        to the challenge during authentication
+        Parameters sent:
+            * serial
+            * nonce (which is the challenge)
+            * signature (which is the signed nonce)
+
+
+        :param request: The Flask request
+        :param g: The Flask global object g
+        :return: dictionary
+        """
+        from privacyidea.lib.token import get_tokens
+        from privacyidea.lib.utils import prepare_result
+        details = {}
+        result = False
+        serial = getParam(request.all_data, "serial", optional=False)
+
+        if serial and "fbtoken" in request.all_data and "pubkey" in request.all_data:
```

Comment:
We need to check that the token with serial ``serial`` is actually a push token. Otherwise, the push API endpoint may be used to update non-push tokens without authentication.

Reply:
done in afca29fe